src/mip/mip_logger.h

```diff
@@ -0,0 +1,200 @@
+    if (mip_logger_get_callback() != NULL && mip_logger_get_level() >= level) \
+        mip_logger_get_callback()(context, mip_logger_get_user_data(), level, fmt, __VA_ARGS__)
+#else
+#define mip_logger_log(context, level, fmt, ...) do {} while (1)
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at fatal level
+///
+///@param context Context of what called this function. Could be a MIP device, serial connection, etc.
+///@param fmt     printf style format string
+///@param ...     Variadic args used to populate the fmt string
+///
+#define mip_logger_fatal(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_FATAL, fmt, __VA_ARGS__)
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at error level
+///@copydetails mip::C::mip_logger_fatal
+#define mip_logger_error(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_ERROR, fmt, __VA_ARGS__)
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at warn level
+///@copydetails mip::C::mip_logger_fatal
+#define mip_logger_warn(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_WARN, fmt, __VA_ARGS__)
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at info level
+///@copydetails mip::C::mip_logger_fatal
+#define mip_logger_info(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_INFO, fmt, __VA_ARGS__)
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at debug level
+///@copydetails mip::C::mip_logger_fatal
+#define mip_logger_debug(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_DEBUG, fmt, __VA_ARGS__)
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at trace level
+///@copydetails mip::C::mip_logger_fatal
+#define mip_logger_trace(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_TRACE, fmt, __VA_ARGS__)
+
+///@}
+///@}
+////////////////////////////////////////////////////////////////////////////////
+
+#ifdef __cplusplus
+} // extern "C"
+} // namespace C
+
+////////////////////////////////////////////////////////////////////////////////
+///@addtogroup mip_cpp
+///@{
+////////////////////////////////////////////////////////////////////////////////
+///@defgroup mip_logger_cpp  Mip Logging [C++]
+///
+///@brief High-level C++ functions for logging information from within the MIP SDK
+///
+/// This module contains functions that allow the MIP SDK to log information
+/// and allows users to override the logging functions
+///
+///@{
+
```

Comment:
Needs "namespace mip {" here.

Reply:
This should still be inside the namespace mip because of the first `namespace mip {` block at the start of the file. I think that is how the other files do it, is that not correct?

Reply:
Ah sorry I missed it, you're right.